src/hooks/useSelector.js

```diff
@@ -53,10 +53,15 @@ function useSelectorWithStoreAndSubscription(
 
   useIsomorphicLayoutEffect(() => {
     function checkForUpdates() {
+      const newReduxState = store.getState()
       try {
-        const newSelectedState = latestSelector.current(store.getState())
+        const newSelectedState = latestSelector.current(newReduxState)
 
         if (equalityFn(newSelectedState, latestSelectedState.current)) {
+          setReduxState(prev => {
+            prev.value = newReduxState
```

Comment:
@salvoravida As I said in other thread, I'm not confident this works for the future. Do you ever feel like just trying my previous snippet???

Reply:
For what i have understand, useRef is not CM safe. its a shared mem between branches. we need an useStateRef, a way to save data to current branch states queue.
I think it is safe, just says React to not re-render, but store a value on that branch.

i dont' think that reading value from the future is ok in CM and event streaming, but i will try your idea asap. 
Do you have tried?

Reply:
This line is mutating the existing state value, though, which is definitely _not_ CM-safe.

Reply:
no, because it is executed only on next update, before next render.

Reply:
on every branch, so every branch has it's own reduxState queue changes. that's why there is NO more tearing, even if with last test, that simulate this edge case.

Reply:
Mmm... I'm going to need to see some examples and take some time to understand why this may be a good idea.  Why would we not want to just spread the state object here?

Reply:
Also, while I don't know how all the internals of React are implemented, what makes you say that every branch has a separate state object?  Can you point to where you're seeing that in the React source?

Reply:
becouse we do not want a re-render as reduxState changes, but selectedState NO.

Reply:
but we also want to Stream "all" redux changes to iternaal component states (that is handled bu react) without missing ones, or without reading from the future, because of that.

Reply:
i think that we no more need nested subscriptions. "100%" queque of reduxChanges streamed to components, means that react can "branches" states when need, every comp has his own queue

Reply:
I'm really not convinced this is a technically correct approach.  I'll have to look at it more later.

Reply:
Also, nesting subscriptions is specifically for ensuring that a component renders with the latest props from its parent, it doesn't have anything to do with CM or Suspense.  That's still necessary.

Reply:
instead i think that the component should not "jump" renders for that case, but process his queue of reduxChanges that will fix slate pros on next render. (but only 1 commit due to batches)

this way we avoid reading from future that is bad in CM. 
But yeah, do some tests by your own .)

Reply:
@salvoravida One note: As I understand, by mutating state in callback that redux can call at any time, what you are doing now is equivalent to reading store.getState(). So, I don't think you avoid reading from future. (If I understand "read from future" correctly.)

Reply:
No. That's like Redis queue FIFO. On ReduxState Changes, we "capture the state" (note where we read the state) and enqueue to the component, that will evaluete it at any time, but "inqueue".

Note2: useState.setState === useReducer.Dispatch Action. read ReactInternals :D

Reply:
In few words: we "broadcast" reduxChanges to React branches. OneToMany.

Reply:
That was my bad. It's not updating it in callback, but in the updater function.

Reply:
I don't understand what this does 🙈 
I removed this and:
```javascript
          setReduxState(prev => {
            prev.value = newReduxState
            return prev
          })
```
I tried it locally, still passed the tests. Am I missing something?

At first I though you were using it as a custom `ref` that can optionally cause a re-render but I'm confused by this:
```
In few words: we "broadcast" reduxChanges to React branches. OneToMany, without lost "frames" !
```
Maybe `reduxState` can be "branched", but the selected state is still a ref. Is that the point?

Reply:
Have you run CM tests? on /test/cm. on SyncRender there is 1 queue only so it's irrilevant.

Every Branch has it's own States queue. so states are Branch safe, Ref are shared beetween renders braches! the are tecnicaly "this" on the component instances.
States are Copied and enqueued on every branched that has different speed and timeline of execution! so the are branch safe.

Reply:
this is what i call statesRef, here are an implementation.
```jsx
import { useState, useCallback } from 'react';

export const useStateRef = initialState => {
  const [state, setState] = useState(() => ({
    current: typeof initialState === 'function' ? initialState() : initialState
  }));

  const setStateRef = useCallback(
    (setter, forceRender) => {
      setState(prev => {
        const current = typeof setter === 'function' ? setter(prev.current) : setter;
        if (forceRender) return { current };
        prev.current = current;
        return prev;
      });
    },
    [setState]
  );
  return [state, setStateRef];
};
```
As CM becomes the standard, i wish React will have a native useStateRef.
It will help libraries stay in sync with React branches.
Maybe i should open an RFC?

Reply:
Yes, I removed those 3 lines, run build, then jest:
```javascript
 PASS  __tests__/all_spec.js (13.458s)
  react-redux
    ✓ check1: updated properly (8836ms)
    ✓ check2: no tearing during update (2ms)
    ✓ check3: ability to interrupt render
    ✓ check4: proper update after interrupt (2406ms)
```
That's why it's difficult to reason about.

Reply:
have you update to last @dai-shi tests? i do not think so. get a pull :D

Reply:
https://github.com/salvoravida/react-redux/commits/fix-use-selector-cm

Reply:
@dai-shi can confirm?

Reply:
`Already up to date.`
I cloned the repo just about 30min ago, so I'm sure it's up to date 😅

Reply:
i will check. there should be something wrong on copying daishi test from his project. that fix was for his new test with.

Reply:
https://github.com/dai-shi/will-this-react-global-state-work-in-concurrent-mode/issues/3#issuecomment-552300031

Reply:
hm, so it's more or less like...:
```javascript
const [refState, forceUpdateRef] = React.useState({ current: null })
React.useImperativeHandle(refState, () => 'some value', [ ...optional_deps ])
```
Isn't this just like a short cut?, I mean, if keeping state in ref and just using `forceUpdate`. If it's within a transition, `forceUpdate` will be scheduled for later but the ref will still change which I think this does. Using `useState` for refs is just like merging a ref with forceUpdate into one line of code:
```javascript
const ref = useRef()
const forceUpdate = useReducer(c => c + 1, 0)[1]
somewhere {
   ref.current = someValue
   forceUpdate()
}
...
const [ref, forceUpdate] = useState({ current: null })
somewhere {
   ref.current = someValue
   forceUpdate({ ...ref })
}
```
Both `forceUpdate` will schedule "for later" if within transition.

Reply:
Yes but using ref, another rendering brahch could update the same value before or after, like shared mem between 3d. Instead states ref are duplicated on every branch and every state ref updater will work on its own branch rendering.

Reply:
Idk, this is how I see it:
```
queue no transition => [
  setReduxState(p => { p.value = reduxState 1; return p }),
  setReduxState(p => { p.value = reduxState 2; return p }), // before p.value is reduxState 1
  setReduxState(p => { p.value = reduxState 3; return p }), // before p.value is reduxState 2
]
start transition queue => [
  later setReduxState(p => { p.value = reduxState 4; return p }), // previous 'p' is unknown
]
queue outside transition => [
  setReduxState(p => { p.value = reduxState 5; return p }), // before p.value is reduxState 3
]
stop transition queue => [
  run previous:
    initial state: reduxState 5 <<<< this is previous 'p' now
    setReduxState(p => { p.value = reduxState 4; return p }), // <<< inconsistent state
]
```
Now, current @dai-shi tests using transition are not exactly right. It's testing dispatching in transition.. but, since there is no Suspense, no component is waiting for anything, so there is no need to queue for later.

Sometimes there may be some flickering on screen, when you use transition but nothing causes interrupting of rendering (like throw promise on Suspense), I've seen on some demos I was testing with, I think this is a bug related to what Dan mentioned on twitter, maybe? Point is, you **can't** test using transition without causing interrupt on render (throw on Suspense), otherwise it is pointless to wrap it in `startTransition`.

Reply:
I'm not following the discussion in deep, but:

> without causing interrupt on render (throw on Suspense)

My understanding is a bit different. We can cause interrupt with heavy computation without throwing promise. That's the whole point of my tests.

P.S. Did I ever test using transition? Just today, I made new checks using `useTransition` based on @eddyw 's demo. No Suspense. No throwing promises. Is it pointless?

Reply:
Never mind, ignore the above 🙈 
I've copied the tests over this PR locally which did have `useTransition` but you finished committing later, so I had them broken